core/node/state_keeper/src/v26_utils.rs

```diff
@@ -0,0 +1,481 @@
+    if expected_asset_id != used_asset_id {
+        return None;
+    }
+
+    Some(origin_token_address)
+}
+
+async fn calculate_expected_token_address(
+    storage: &mut Connection<'_, Core>,
+    l2_legacy_shared_bridge_address: Address,
+    l1_token_address: Address,
+) -> anyhow::Result<Address> {
+    // The source of truth for this logic is `L2SharedBridgeLegacy._calculateCreate2TokenAddress`
+
+    let beacon_proxy_bytecode_hash_key = StorageKey::new(
+        AccountTreeId::new(l2_legacy_shared_bridge_address),
+        L2_LEGACY_SHARED_BRIDGE_BEACON_PROXY_BYTECODE_KEY,
+    );
+    let beacon_proxy_bytecode_hash = storage
+        .storage_web3_dal()
+        .get_value(&beacon_proxy_bytecode_hash_key)
+        .await?;
+
+    let beacon_address_key = StorageKey::new(
+        AccountTreeId::new(l2_legacy_shared_bridge_address),
+        L2_LEGACY_SHARED_BRIDGE_UPGRADEABLE_BEACON_ADDRESS_KEY,
+    );
+    let beacon_address = storage
+        .storage_web3_dal()
+        .get_value(&beacon_address_key)
+        .await?;
+    let beacon_address = h256_to_address(&beacon_address);
+
+    let params = Create2DeploymentParams {
+        salt: address_to_h256(&l1_token_address),
+        bytecode_hash: beacon_proxy_bytecode_hash,
+        raw_constructor_input: ethabi::encode(&[
+            Token::Address(beacon_address),
+            Token::Bytes(vec![]),
+        ]),
+    };
+
+    Ok(params.derive_address(l2_legacy_shared_bridge_address))
+}
+
+#[derive(Debug)]
+enum LegacyTokenStatus {
+    // In case the deposited token is either not legacy or registered.
+    NotLegacyOrRegistered,
+    // In case the deposited token is a legacy token that has not
+    // been registered
+    Legacy,
+    // This is the case for an unexpected state where the predicted l1
+    // address does not match the storage one. It should never happen, but
+    // we return an error instead of a panic for increased liveness / easier debugging,
+    // while the transaction will still not be allowed to pass through.
+    UnexpectedDifferentL1Address(Address, Address),
+}
+
+/// Checks whether the token is legacy. It is legacy if both of the
+/// following is true:
+/// - It is present in legacy shared bridge
+/// - It is not present in the L2 native token vault
+async fn is_l2_token_legacy(
+    storage: &mut Connection<'_, Core>,
+    l2_legacy_shared_bridge_address: Address,
+    l2_token_address: Address,
+    expected_l1_address: Address,
+) -> anyhow::Result<LegacyTokenStatus> {
+    // 1. Read l1 token address from L2 shared bridge (must not be 0)
+
+    let stored_l1_address_key = StorageKey::new(
+        AccountTreeId::new(l2_legacy_shared_bridge_address),
+        get_address_mapping_key(&l2_token_address, L2_LEGACY_SHARED_BRIDGE_L1_ADDRESSES_KEY),
+    );
+    let stored_l1_address = storage
+        .storage_web3_dal()
+        .get_value(&stored_l1_address_key)
```

Comment:
Here and in other `get_value()` calls: Does it matter if new storage logs are inserted in the middle of processing? In this case, later `get_value()` calls may read newer data. For a consistent storage view, it could be possible to use `get_historical_value_unchecked()` instead with a block number obtained at the start of processing. AFAIU, Postgres transactions with the default ["read committed" isolation level](https://www.postgresql.org/docs/14/transaction-iso.html#XACT-READ-COMMITTED) would still see new storage logs, and using stronger isolation could harm performance, so having such app-level consistency looks more appropriate.

Reply:
I could use `get_historical_value_unchecked`, but in the end of the day it does not seem like it matters too much. If we did not read a newer value, they will be read eventually during the next iterations, the current approach seems to incur less code

Reply:
Okay, let's use the `get_historical_value_unchecked` just in case